{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Day 2\n",
    "\n",
    "Today, we will start using nf-core pipelines to find differentially abundant genes in our dataset. \n",
    "We are using data from the following paper: https://www.nature.com/articles/s41593-023-01350-3#Sec10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "1. Please take some time to read through the paper and understand their approach, hypotheses and goals."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "What was the objective of the study?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- The study's objective was to investigate how patients with chronic neuropathic pain react to opioid administration and withdrawal. Using a mouse model of spared nerve injury (SNI), the authors administered oxycodone and compared gene expression changes with control (sham) and non-opioid groups."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "What do the conditions mean?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- oxy: mice that were given oxycodone \n",
    "\n",
    "\n",
    "- sal: mice that were given a saline (control group)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What do the genotypes mean?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- SNI: spared nerve injury\n",
    "\n",
    "\n",
    "- Sham: healthy / control group (which also had surgery, but no chronic pain)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Imagine you are the bioinformatician in the group who conducted this study. They hand you the raw files and ask you to analyze them.\n",
    "\n",
    "What would you do?    \n",
    "- The aim would be to perform a differential gene expression analysis. For this, the pipeline we looked at yesterday could be suitable or at least a starting point.\n",
    "- This approach will hopefully allow to identify transcriptional changes associated with neuropathic pain, opioid exposure and withdrawal\n",
    "\n",
    "Which groups would you compare to each other?  \n",
    "- SNI-Oxy vs. Sham-Oxy to find out what differs between SNI / Sham genotypes with drug treatment  \n",
    "- SNI-Oxy vs. SNI-sal to find out the genetic effects of the drug treatment in SNI mice  \n",
    "- Sham-Oxy vs. Sham-sal to find out the genetic effects of the drug treatment in Sham mice (according to the paper, research on this has already been done)  \n",
    "  \n",
    "Please also mention which outcome you would expect to see from each comparison.  \n",
    "- 1) a number of differntially expressed genes that are involved in the different reactions to drug treatment and withdrawal in SNI / Sham conditions  \n",
    "- 2) a number of differentially expressed genes linked to the organism's transcriptional response to oxycodone in the SNI condition    \n",
    "- 3) a number of differentially expressed genes linked to the organism's transcriptional response to oxycodone without SNI  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Your group gave you a very suboptimal excel sheet (conditions_runs_oxy_project.xlsx) to get the information you need for each run they uploaded to the SRA.<br>\n",
    "So, instead of directly diving into downloading the data and starting the analysis, you first need to sort the lazy table.<br>\n",
    "Use Python and Pandas to get the table into a more sensible order.<br>\n",
    "Then, perform some overview analysis and plot the results\n",
    "1. How many samples do you have per condition?\n",
    "2. How many samples do you have per genotype?\n",
    "3. How often do you have each condition per genotype?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samples per condition:\n",
      "Oxy    8\n",
      "Sal    8\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "conditions_table = (\"conditions_runs_oxy_project.xlsx\")\n",
    "\n",
    "# with \"run\" as index column: \n",
    "df = pd.read_excel(conditions_table, index_col=\"Run\")\n",
    "\n",
    "# other ideas that were discussed: \n",
    "# df.fillna(False) \n",
    "# df.replace(\"x\",True)\n",
    "# or: one condition column with \"sal\" / \"oxy\"\n",
    "\n",
    "# 1. How many samples do you have per condition?\n",
    "condition_counts = pd.Series({\n",
    "    \"Oxy\": (df[\"Condition: Oxy\"] == \"x\").sum(),\n",
    "    \"Sal\": (df[\"condition: Sal\"] == \"x\").sum()\n",
    "})\n",
    "\n",
    "\n",
    "print(\"Samples per condition:\")\n",
    "print(condition_counts.to_string())\n",
    "\n",
    "# other idea: 0 and 1, then count\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samples per genotype:\n",
      "SNI     8\n",
      "Sham    8\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# 2. How many samples do you have per genotype?\n",
    "\n",
    "genotype_counts = pd.Series({\n",
    "    \"SNI\": (df[\"Genotype: SNI\"] == \"x\").sum(),\n",
    "    \"Sham\": (df[\"Genotype: Sham\"] == \"x\").sum()\n",
    "})\n",
    "\n",
    "print(\"Samples per genotype:\")\n",
    "print(genotype_counts.to_string())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Condition per genotype:\n",
      "      Sal  Oxy\n",
      "Sham    4    4\n",
      "SNI     4    4\n"
     ]
    }
   ],
   "source": [
    "# 3. How often do you have each condition per genotype?\n",
    "cond_per_geno = pd.crosstab(\n",
    "    df[\"Genotype: SNI\"] == \"x\",\n",
    "    df[\"Condition: Oxy\"] == \"x\"\n",
    ")\n",
    "cond_per_geno.index = [\"Sham\", \"SNI\"] \n",
    "cond_per_geno.columns = [\"Sal\", \"Oxy\"]\n",
    "\n",
    "print(\"Condition per genotype:\")\n",
    "print(cond_per_geno)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "They were so kind to also provide you with the information of the number of bases per run, so that you can know how much space the data will take on your Cluster.<br>\n",
    "Add a new column to your fancy table with this information (base_counts.csv) and sort your dataframe according to this information and the condition.\n",
    "\n",
    "Then select the 2 smallest runs from your dataset and download them from SRA (maybe an nf-core pipeline can help here?...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>RNA-seq</th>\n",
       "      <th>DNA-seq</th>\n",
       "      <th>condition: Sal</th>\n",
       "      <th>Condition: Oxy</th>\n",
       "      <th>Genotype: SNI</th>\n",
       "      <th>Genotype: Sham</th>\n",
       "      <th>Bases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR23195505</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6922564500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195506</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>7859530800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195507</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>8063298900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195508</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6927786900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195509</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7003550100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195510</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7377388500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195511</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>6456390900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195512</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>7462857900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195513</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8099181600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195514</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>7226808600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195515</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>8169101700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195516</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6203117700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195517</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6863840400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195518</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7908500400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195519</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>6996050100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195520</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>7858146000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Patient RNA-seq  DNA-seq condition: Sal Condition: Oxy  \\\n",
       "Run                                                                  \n",
       "SRR23195505       ?       x      NaN              x            NaN   \n",
       "SRR23195506       ?       x      NaN            NaN              x   \n",
       "SRR23195507       ?       x      NaN              x            NaN   \n",
       "SRR23195508       ?       x      NaN            NaN              x   \n",
       "SRR23195509       ?       x      NaN            NaN              x   \n",
       "SRR23195510       ?       x      NaN              x            NaN   \n",
       "SRR23195511       ?       x      NaN            NaN              x   \n",
       "SRR23195512       ?       x      NaN              x            NaN   \n",
       "SRR23195513       ?       x      NaN              x            NaN   \n",
       "SRR23195514       ?       x      NaN            NaN              x   \n",
       "SRR23195515       ?       x      NaN              x            NaN   \n",
       "SRR23195516       ?       x      NaN            NaN              x   \n",
       "SRR23195517       ?       x      NaN            NaN              x   \n",
       "SRR23195518       ?       x      NaN              x            NaN   \n",
       "SRR23195519       ?       x      NaN            NaN              x   \n",
       "SRR23195520       ?       x      NaN              x            NaN   \n",
       "\n",
       "            Genotype: SNI Genotype: Sham       Bases  \n",
       "Run                                                   \n",
       "SRR23195505             x            NaN  6922564500  \n",
       "SRR23195506           NaN              x  7859530800  \n",
       "SRR23195507           NaN              x  8063298900  \n",
       "SRR23195508             x            NaN  6927786900  \n",
       "SRR23195509             x            NaN  7003550100  \n",
       "SRR23195510             x            NaN  7377388500  \n",
       "SRR23195511           NaN              x  6456390900  \n",
       "SRR23195512           NaN              x  7462857900  \n",
       "SRR23195513             x            NaN  8099181600  \n",
       "SRR23195514           NaN              x  7226808600  \n",
       "SRR23195515           NaN              x  8169101700  \n",
       "SRR23195516             x            NaN  6203117700  \n",
       "SRR23195517             x            NaN  6863840400  \n",
       "SRR23195518             x            NaN  7908500400  \n",
       "SRR23195519           NaN              x  6996050100  \n",
       "SRR23195520           NaN              x  7858146000  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_counts = \"base_counts.csv\"\n",
    "\n",
    "bases = pd.read_csv(base_counts, index_col=\"Run\")\n",
    "\n",
    "# merge based on \"Run\" column\n",
    "df = df.merge(bases, on=\"Run\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>RNA-seq</th>\n",
       "      <th>DNA-seq</th>\n",
       "      <th>condition: Sal</th>\n",
       "      <th>Condition: Oxy</th>\n",
       "      <th>Genotype: SNI</th>\n",
       "      <th>Genotype: Sham</th>\n",
       "      <th>Bases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR23195516</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6203117700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195511</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>6456390900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195517</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6863840400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195505</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6922564500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195508</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6927786900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195519</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>6996050100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195509</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7003550100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195514</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>7226808600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195510</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7377388500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195512</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>7462857900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195520</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>7858146000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195506</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>7859530800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195518</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7908500400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195507</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>8063298900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195513</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8099181600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23195515</th>\n",
       "      <td>?</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>8169101700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Patient RNA-seq  DNA-seq condition: Sal Condition: Oxy  \\\n",
       "Run                                                                  \n",
       "SRR23195516       ?       x      NaN            NaN              x   \n",
       "SRR23195511       ?       x      NaN            NaN              x   \n",
       "SRR23195517       ?       x      NaN            NaN              x   \n",
       "SRR23195505       ?       x      NaN              x            NaN   \n",
       "SRR23195508       ?       x      NaN            NaN              x   \n",
       "SRR23195519       ?       x      NaN            NaN              x   \n",
       "SRR23195509       ?       x      NaN            NaN              x   \n",
       "SRR23195514       ?       x      NaN            NaN              x   \n",
       "SRR23195510       ?       x      NaN              x            NaN   \n",
       "SRR23195512       ?       x      NaN              x            NaN   \n",
       "SRR23195520       ?       x      NaN              x            NaN   \n",
       "SRR23195506       ?       x      NaN            NaN              x   \n",
       "SRR23195518       ?       x      NaN              x            NaN   \n",
       "SRR23195507       ?       x      NaN              x            NaN   \n",
       "SRR23195513       ?       x      NaN              x            NaN   \n",
       "SRR23195515       ?       x      NaN              x            NaN   \n",
       "\n",
       "            Genotype: SNI Genotype: Sham       Bases  \n",
       "Run                                                   \n",
       "SRR23195516             x            NaN  6203117700  \n",
       "SRR23195511           NaN              x  6456390900  \n",
       "SRR23195517             x            NaN  6863840400  \n",
       "SRR23195505             x            NaN  6922564500  \n",
       "SRR23195508             x            NaN  6927786900  \n",
       "SRR23195519           NaN              x  6996050100  \n",
       "SRR23195509             x            NaN  7003550100  \n",
       "SRR23195514           NaN              x  7226808600  \n",
       "SRR23195510             x            NaN  7377388500  \n",
       "SRR23195512           NaN              x  7462857900  \n",
       "SRR23195520           NaN              x  7858146000  \n",
       "SRR23195506           NaN              x  7859530800  \n",
       "SRR23195518             x            NaN  7908500400  \n",
       "SRR23195507           NaN              x  8063298900  \n",
       "SRR23195513             x            NaN  8099181600  \n",
       "SRR23195515           NaN              x  8169101700  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted = df.sort_values(by=\"Bases\", ascending=1)\n",
    "sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fetchngs pipeline\n",
    "\n",
    "!nextflow run nf-core/fetchngs --input computational-workflows-2025/notebooks/day_02/ids.csv -profile docker --outdir day02/fetchngs_out --max_memory \"6GB\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While your files are downloading, get back to the paper and explain how you would try to reproduce the analysis.<br>\n",
    "\n",
    "The paper did a differential gene analysis on the following pairs:\n",
    "- SNI-Oxy versus Sham-Sal\n",
    "- SNI-Sal versus Sham-Sal\n",
    "- Sham-Oxy versus Sham-Sal\n",
    "I would try to reproduce the analysis as described in the \"Bioninformatics analysis\" part of the Methods section.   \n",
    "The method describes:  \n",
    "- Read alignment and DGE with HISAT2, HTSeq and DSeq2\n",
    "- creating plots\n",
    "- analysis of enriched pathways with IPA\n",
    "\n",
    "Since this section does not state any detais on how the analysis was performed, which parameters were used etc, and does not link to a github or similar, it could become difficult to reproduce the analysis exactly.\n",
    "\n",
    "\n",
    "When you are done with this shout, so we can discuss the different ideas."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "workflows_SS25_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
